default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1573 @@
+                        if old_part.partClass in part_class:
+                            if new_part.capacity >= old_part.capacity:
+                                a = new_part
+                                b = old_part
+                            else:
+                                a = old_part
+                                b = new_part
+                            if (self.production_cost[pid][a.name] <= self.production_cost[pid][b.name]
+                                    and {x for x in a.mountableSlotTypes} >= {x for x in b.mountableSlotTypes}):
+                                # ToDo: add production_time as additional condition?
+                                self.strictly_worse_parts[a.name].append(b.name)
+                    break
+        available_parts = sorted(self.strictly_worse_parts.keys(),
+                                 key=lambda item: _get_part_type(item).capacity, reverse=True)
+
+        # in case of a load, we need to rebuild our Cache.
+        if not self.testhulls:
+            print "Testhull cache not found. This may happen only at first turn after game start or load."
+            for hullname in available_hulls:
+                des = [des for des in testdesign_names_part if des.endswith(hullname)]
+                if des:
+                    self.testhulls.add(hullname)
+            if verbose:
+                print "Rebuilt Cache. The following hulls are used in testdesigns for parts: ", self.testhulls
+
+        # 1. Update hull test designs
+        print "Updating Testdesigns for hulls..."
+        if verbose:
+            print "Available Hulls: ", available_hulls
+            print "Existing Designs (prefix: %s): " % TESTDESIGN_NAME_HULL,
+            print [x.replace(TESTDESIGN_NAME_HULL, "") for x in testdesign_names_hull]
+        for hull in [get_hulltype(hullname) for hullname in available_hulls
+                     if "%s_%s" % (TESTDESIGN_NAME_HULL, hullname) not in testdesign_names_hull]:
+            partlist = len(hull.slots) * [""]
+            testdesign_name = "%s_%s" % (TESTDESIGN_NAME_HULL, hull.name)
+            res = fo.issueCreateShipDesignOrder(testdesign_name, "TESTPURPOSE ONLY", hull.name,
+                                                partlist, "", "fighter", False)
+            if res:
+                print "Success: Added Test Design %s, with result %d" % (testdesign_name, res)
+            else:
+                print "Error: When adding test design %s - got result %d but expected 1" % (testdesign_name, res)
+                continue
+
+        # 2. Cache the list of buildable ship hulls for each planet
+        print "Caching buildable hulls per planet..."
+        testname = "%s_%s" % (TESTDESIGN_NAME_HULL, "%s")
+        for pid in planets_with_shipyards:
+            self.hulls_for_planets[pid] = []
+        for hullname in available_hulls:
+            testdesign = _get_design_by_name(testname % hullname)
+            try:
```

Comment:
You expect `AttributeError` because `ship_design` can be null. Can you remove try catch and make special check for this case.  It will be more readable.